### DISCUSSION TITLE
How can I convert the JAX Tracer object (<class 'jax.interpreters.batching.BatchTracer'>) to numpy.ndarray?

### DISCUSSION THREAD

**Discussion by XuHu0529** [jax-ml/jax#13488] [Q&A]
I am a newer in JAX!!
When I use `jax.vmap` function, I get a type of `<jax.interpreters.batching.BatchTracer>` for every array. Because I am using a function which is under numpy setting, I have to convert the tensor/array to numpy.ndarray type. I try to use `np.asarray()`or `jax.device_get()`but it doesn't work any more.
Is there any other method to help me with that?

**Comment by jakevdp** [comment]
*Edit: there is now a FAQ entry on this topic: https://jax.readthedocs.io/en/latest/faq.html#how-can-i-convert-a-jax-tracer-to-a-numpy-array*

---

Hi, thanks for the question. This comes up frequently enough that I'm probably going to add it to the FAQ. See if the answer in https://github.com/google/jax/discussions/13295 helps - if not, then feel free to post back here and I'll try to help with your specific problem. But in general, you can only use JAX transformations like `jit` and `vmap` on JAX functions, not on functions that require numpy arrays as inputs.
